cpp/src/hash/md5_hash.cu

```diff
@@ -27,46 +31,265 @@
+    while (len > md5_chunk_size + copylen) {
+      memcpy(hash_state->buffer, data + copylen, md5_chunk_size);
+      md5_hash_step(hash_state);
+      copylen += md5_chunk_size;
+    }
+
+    // The remaining data chunk does not fill the buffer. We copy the data into
+    // the buffer but do not trigger a hash step yet.
+    memcpy(hash_state->buffer, data + copylen, len - copylen);
+    hash_state->buffer_length = len - copylen;
+  }
+}
+
+/**
+ * @brief MD5 typed element processor.
+ *
+ * This accepts typed data, normalizes it, and performs processing on raw bytes.
+ */
+template <typename T>
+void CUDA_DEVICE_CALLABLE md5_process(T const& key, md5_intermediate_data* hash_state)
+{
+  if constexpr (is_fixed_width<T>() && !is_chrono<T>()) {
+    if constexpr (is_floating_point<T>()) {
+      T const normalized_key = normalize_nans_and_zeros_helper<T>(key);
+      uint8_t const* data    = reinterpret_cast<uint8_t const*>(&normalized_key);
+      uint32_t constexpr len = sizeof(T);
+      md5_process_bytes(data, len, hash_state);
+    } else {
+      uint8_t const* data    = reinterpret_cast<uint8_t const*>(&key);
+      uint32_t constexpr len = sizeof(T);
+      md5_process_bytes(data, len, hash_state);
+    }
+  } else if constexpr (std::is_same_v<T, string_view>) {
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(key.data());
+    uint32_t len        = static_cast<uint32_t>(key.size_bytes());
+    md5_process_bytes(data, len, hash_state);
+  } else {
+    cudf_assert(false && "Unsupported type for hash function.");
+  }
+}
 
 // MD5 supported leaf data type check
 bool md5_type_check(data_type dt)
 {
   return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
 }
 
-}  // namespace
+struct MD5ListHasher {
+  template <typename T,
+            CUDF_ENABLE_IF((is_fixed_width<T>() && !is_chrono<T>()) ||
+                           std::is_same_v<T, string_view>)>
+  void CUDA_DEVICE_CALLABLE operator()(column_device_view data_col,
+                                       size_type offset_begin,
+                                       size_type offset_end,
+                                       md5_intermediate_data* hash_state) const
+  {
+    for (size_type i = offset_begin; i < offset_end; i++) {
+      if (data_col.is_valid(i)) { md5_process(data_col.element<T>(i), hash_state); }
+    }
+  }
 
-namespace detail {
+  template <typename T,
+            CUDF_ENABLE_IF((!is_fixed_width<T>() || is_chrono<T>()) &&
+                           !std::is_same_v<T, string_view>)>
+  void CUDA_DEVICE_CALLABLE
+  operator()(column_device_view, size_type, size_type, md5_intermediate_data*) const
+  {
+    cudf_assert(false && "Unsupported type for hash function.");
+  }
+};
+
+struct MD5Hash {
+  void CUDA_DEVICE_CALLABLE finalize(md5_intermediate_data* hash_state, char* result_location) const
+  {
+    auto const full_length = (static_cast<uint64_t>(hash_state->message_length)) << 3;
```

Comment:
The compiler will do the same shift optimization for you and this is infinitely more readable. 
```suggestion
    auto const full_length = hash_state->message_length * 8;
```

Reply:
This variable isn't used until much further below. In general, keep declarations as close to where they are used as possible.